File: Container With Most Water.py
from typing import List

class Solution:
    def maxAreaSlow(self, height: List[int]) -> int:
        highest = float('-inf')
        for i in range(len(height)):
            for j in range(i, len(height)):
                highest = max(highest, (j-i)*min(height[i], height[j]))
        return highest

    def maxAreaOptimised(self, height: List[int]) -> int:
        highest = 0
        l = 0
        r = len(height) - 1
        while l < r:
            highest = max(highest, (r-l)*min(height[r], height[l]))
            if height[l] < height[r]:
                l += 1
            else:
                r -= 1
        return highest

File: cpp/Container With Most Water.cpp
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;


class Solution {
public:
    int maxArea(vector<int>& height) {
        int l = 0;
        int r = height.size() - 1;

        int max_water = 0;

        while (l<r) {
            max_water = max(max_water, (r-l)* min(height[l], height[r]));

            if (height[l] < height[r]) {
                l++;
            } else {
                r--;
            }
        }
        return max_water;      
    }
};